### PR TITLE
added shutdownevent for graceful shutdown

### PR DESCRIPTION
**Pull request by haomingzhang**

### CODE REVIEW

File: events/event_manager.go [hunk 1]
```diff
@@ -13,7 +13,8 @@ type eventManager struct {
 
 func (em *eventManager) Emit(selector apid.EventSelector, event apid.Event) {
 	log.Debugf("emit selector: '%s' event %v: %v", selector, &event, event)
-	if !em.dispatchers[selector].Send(event) {
+	eventDispatcher := em.dispatchers[selector]
```
Comment (theganyo) [inline events/event_manager.go]: Why change this?

File: apid.go [hunk 2]
```diff
@@ -69,6 +76,24 @@ func InitializePlugins() {
 	log.Debugf("done initializing plugins")
 }
 
+func ShutdownPlugins() {
+	Events().EmitWithCallback(ShutdownEventSelector, ShutdownEvent{"apid is going to shutdown"}, ShutdownHandler)
+}
+
+
+func ShutdownHandler(event Event) {
```
Comment (theganyo) [inline apid.go]: I'd recommend making this an anonymous function inside the ShutdownPlugins() function instead of exposing it.

Comment (haomingzhang) [reply apid.go]: OK, changed to `shutdownHandler`

File: events/event_manager.go (hunk 1 above)
Comment (haomingzhang) [reply events/event_manager.go]: Not sure about it. I noticed that a selector is added to `dispatchers` by `Listen`.
I'm afraid that if you `Emit` to a selector before anyone listens to it, `dispatchers[selector]` can be `nil`.

File: events/events_test.go [hunk 3]
```diff
@@ -287,8 +289,75 @@ var _ = Describe("Events Service", func() {
...
+			Version: "1.0.0",
+			ExtraData: xData,
+		}
+		return
+	}
+	return p
+
+}
+
 type test_handler struct {
```
Comment (sramamoorthy73) [inline events/events_test.go:365]: Other than Scott's comments, looks ok. Now, we have to hook on to analytics plugin.

File: events/event_manager.go (hunk 1 above)
Comment (theganyo) [reply events/event_manager.go]: Ah. This was part of the design. A nil dispatcher is ok. See: 
```
func (d *dispatcher) Send(e apid.Event) bool {
	if d != nil {
		d.channel <- e
		return true
	}
	return false
}
```

File: apid.go (hunk 2 above)
Comment (sramamoorthy73) [reply apid.go]: ShutdownPlugins() should be called when scaffold catches the signal for shutdown. Where have you added that?

File: apid.go [hunk 4]
```diff
@@ -1,9 +1,15 @@
...
+	"errors"
+	"os"
+	"time"
+)
 
 const (
-	SystemEventsSelector EventSelector = "system event"
+	SystemEventsSelector  EventSelector = "system event"
+	ShutdownEventSelector EventSelector = "shutdown event"
+	ShutdownTimeout       int           = 10
```
Comment (theganyo) [inline apid.go]: Why not make this a time.Duration directly?

Comment (haomingzhang) [reply apid.go]: Good idea

File: apid.go [hunk 5]
```diff
@@ -9,7 +9,7 @@ import (
 const (
 	SystemEventsSelector  EventSelector = "system event"
 	ShutdownEventSelector EventSelector = "shutdown event"
-	ShutdownTimeout       int           = 10
+	ShutdownTimeout       time.Duration = time.Duration(10) * time.Second
```
Comment (theganyo) [inline apid.go]: Did you know this could simply be?
`ShutdownTimeout = 10 * time.Second`

Comment (haomingzhang) [reply apid.go]: OK, will change it

**Pull request merged**